### PR TITLE
feature(check_hw_perf): Validate hw performance for performance test

### PR DESCRIPTION
**Pull request by aleksbykov**
Run and check results of hw performance test before usual performance
test, to be sure that all db nodes have similar performance on hardware

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by KnifeyMoloko** [comment]
In general LGTM. 
Note: The current acceptable margin values for the benchmarks we perform are basically my arbitrary decision, so they might be to strict and we end up with jobs failing because one or two nodes fail to be "within margin" with their benchmark performance. The upside for that is we should have more confidence about the performance results - if we see a degradation, it's less likely to be caused by a single faulty node.

We might want to mitigate the risk of failing jobs. Some ideas:

1) set the margin values in the test yaml, so that we can decide how strict / lenient we want the checks to be
2) rather than raising an error log message, we might want to do it as a warning and add the table with the node benchmarks to the test email. That way if we see a degradation, we can check if some of the nodes had performance issues when benchmarked
3) replace the faulty node before the test starts if we see it's performance varies from the others and then rerun the benchmark just on that node

**Comment by KnifeyMoloko** [comment]
> @KnifeyMoloko I'm trying to understand somehting (which I probably forgot since you've done this task):
> 
> Here:
> 
> ```
> 16:40:26  < t:2022-11-22 09:40:22,375 f:performance_regression_test.py l:54   c:PerformanceRegressionTest p:INFO  > Validate node benchmarks results
> 16:40:26  < t:2022-11-22 09:40:22,375 f:performance_regression_test.py l:61   c:PerformanceRegressionTest p:ERROR > HW performance test on node perf-regression-ubuntu-db-node-fa37917a-1 has bad results for cassandra_fio_read_bw : {'value': 43551, 'average': 69693.0, 'average_ratio': 0.6248977659162326, 'is_within_margin': False}
> 16:40:26  < t:2022-11-22 09:40:22,375 f:performance_regression_test.py l:61   c:PerformanceRegressionTest p:ERROR > HW performance test on node perf-regression-ubuntu-db-node-fa37917a-2 has bad results for cassandra_fio_read_bw : {'value': 42866, 'average': 69693.0, 'average_ratio': 0.6150689452312283, 'is_within_margin': False}
> 16:40:26  < t:2022-11-22 09:40:22,375 f:performance_regression_test.py l:61   c:PerformanceRegressionTest p:ERROR > HW performance test on node perf-regression-ubuntu-db-node-fa37917a-3 has bad results for cassandra_fio_read_bw : {'value': 43916, 'average': 69693.0, 'average_ratio': 0.6301350207337896, 'is_within_margin': False}
> 16:40:26  < t:2022-11-22 09:40:22,376 f:performance_regression_test.py l:63   c:PerformanceRegressionTest p:CRITICAL > DB Cluster doesn't have equal hw performance result {'perf-regression-ubuntu-db-node-fa37917a-1': {'sysbench_eps': {'value': 5549.82, 'average': 5529.618333333333, 'average_ratio': 1.0036533564251422, 'is_within_margin': True}, 'cassandra_fio_read_bw': {'value': 43551, 'average': 69693.0, 'average_ratio': 0.6248977659162326, 'is_within_margin': False}, 'cassandra_fio_write_bw': {'value': 88988, 'average': 63120.583333333336, 'average_ratio': 1.4098095312279275, 'is_within_margin': True}}, 'perf-regression-ubuntu-db-node-fa37917a-2': {'sysbench_eps': {'value': 5546.21, 'average': 5529.618333333333, 'average_ratio': 1.0030005084739122, 'is_within_margin': True}, 'cassandra_fio_read_bw': {'value': 42866, 'average': 69693.0, 'average_ratio': 0.6150689452312283, 'is_within_margin': False}, 'cassandra_fio_write_bw': {'value': 89687, 'average': 63120.583333333336, 'average_ratio': 1.4208835733721303, 'is_within_margin': True}}, 'perf-regression-ubuntu-db-node-fa37917a-3': {'sysbench_eps': {'value': 5525.48, 'average': 5529.618333333333, 'average_ratio': 0.9992516059728053, 'is_within_margin': True}, 'cassandra_fio_read_bw': {'value': 43916, 'average': 69693.0, 'average_ratio': 0.6301350207337896, 'is_within_margin': False}, 'cassandra_fio_write_bw': {'value': 88670, 'average': 63120.583333333336, 'average_ratio': 1.4047715549734832, 'is_within_margin': True}}}
> ```
> 
> 1. I see values for all the 3 nodes, they look close enough, why did we decided that they aren't within margin?
> 2. What is the average and where do we take it from?

Re: 1 - The "within_margin" check is simply a check if the value for a given node deviates from the average for that node type by a certaiain % amount - the % is currently hardcoded in the benchmaks code.
Re 2: - we save the benchmark results in Elastic (here's a dashboard: https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/92f47acad14bc877780608a928a15647), so we compare against the average for benchmark results for a given node type that we store there.

**Comment by roydahan** [comment]
> > @KnifeyMoloko I'm trying to understand somehting (which I probably forgot since you've done this task):
> > Here:
> > ```
> > 16:40:26  < t:2022-11-22 09:40:22,375 f:performance_regression_test.py l:54   c:PerformanceRegressionTest p:INFO  > Validate node benchmarks results
> > 16:40:26  < t:2022-11-22 09:40:22,375 f:performance_regression_test.py l:61   c:PerformanceRegressionTest p:ERROR > HW performance test on node perf-regression-ubuntu-db-node-fa37917a-1 has bad results for cassandra_fio_read_bw : {'value': 43551, 'average': 69693.0, 'average_ratio': 0.6248977659162326, 'is_within_margin': False}
> > 16:40:26  < t:2022-11-22 09:40:22,375 f:performance_regression_test.py l:61   c:PerformanceRegressionTest p:ERROR > HW performance test on node perf-regression-ubuntu-db-node-fa37917a-2 has bad results for cassandra_fio_read_bw : {'value': 42866, 'average': 69693.0, 'average_ratio': 0.6150689452312283, 'is_within_margin': False}
> > 16:40:26  < t:2022-11-22 09:40:22,375 f:performance_regression_test.py l:61   c:PerformanceRegressionTest p:ERROR > HW performance test on node perf-regression-ubuntu-db-node-fa37917a-3 has bad results for cassandra_fio_read_bw : {'value': 43916, 'average': 69693.0, 'average_ratio': 0.6301350207337896, 'is_within_margin': False}
> > 16:40:26  < t:2022-11-22 09:40:22,376 f:performance_regression_test.py l:63   c:PerformanceRegressionTest p:CRITICAL > DB Cluster doesn't have equal hw performance result {'perf-regression-ubuntu-db-node-fa37917a-1': {'sysbench_eps': {'value': 5549.82, 'average': 5529.618333333333, 'average_ratio': 1.0036533564251422, 'is_within_margin': True}, 'cassandra_fio_read_bw': {'value': 43551, 'average': 69693.0, 'average_ratio': 0.6248977659162326, 'is_within_margin': False}, 'cassandra_fio_write_bw': {'value': 88988, 'average': 63120.583333333336, 'average_ratio': 1.4098095312279275, 'is_within_margin': True}}, 'perf-regression-ubuntu-db-node-fa37917a-2': {'sysbench_eps': {'value': 5546.21, 'average': 5529.618333333333, 'average_ratio': 1.0030005084739122, 'is_within_margin': True}, 'cassandra_fio_read_bw': {'value': 42866, 'average': 69693.0, 'average_ratio': 0.6150689452312283, 'is_within_margin': False}, 'cassandra_fio_write_bw': {'value': 89687, 'average': 63120.583333333336, 'average_ratio': 1.4208835733721303, 'is_within_margin': True}}, 'perf-regression-ubuntu-db-node-fa37917a-3': {'sysbench_eps': {'value': 5525.48, 'average': 5529.618333333333, 'average_ratio': 0.9992516059728053, 'is_within_margin': True}, 'cassandra_fio_read_bw': {'value': 43916, 'average': 69693.0, 'average_ratio': 0.6301350207337896, 'is_within_margin': False}, 'cassandra_fio_write_bw': {'value': 88670, 'average': 63120.583333333336, 'average_ratio': 1.4047715549734832, 'is_within_margin': True}}}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 1. I see values for all the 3 nodes, they look close enough, why did we decided that they aren't within margin?
> > 2. What is the average and where do we take it from?
> 
> @roydahan Re: 1 - The "within_margin" check is simply a check if the value for a given node deviates from the average for that node type by a certaiain % amount - the % is currently hardcoded in the benchmaks code. 

It's weird that we see such a big gap between the "Average for that node type".
Could it be that our query of average of the same node type is incorrect?
Anyway, I think that for now we shouldn't do this comparison and compare the nodes only between themselves.

Re 2: - we save the benchmark results in Elastic (here's a dashboard: https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/92f47acad14bc877780608a928a15647), so we compare against the average for benchmark results for a given node type that we store there.

We should continue recording these results and once we have enough data, we should go back to the numbers and look at the statistics.

**Comment by fruch** [comment]
@aleksbykov 
I've would recommend splitting  c46bca569e766518bcbc2df42b87d297c0cce6ea to it's own PR (once proven working)

**Comment by aleksbykov** [comment]
@fruch @KnifeyMoloko @fruch could you review the PR?

File: configurations/run_db_node_benchmarks.yaml [hunk 1]
```diff
@@ -0,0 +1 @@
+run_db_node_benchmarks: true
```
Comment (roydahan) [inline configurations/run_db_node_benchmarks.yaml:1]: You don't use it anywhere.
Please add it to the Jenkins pipeline of the jobs we want to enable.

Comment (aleksbykov) [reply configurations/run_db_node_benchmarks.yaml:1]: i added to new jenkins file with enabled test for hw benchmark

**Comment by aleksbykov** [comment]
@roydahan @KnifeyMoloko  could you review?

File: performance_regression_test.py [hunk 2]
```diff
@@ -46,7 +46,26 @@ def __init__(self, *args):
...
+        if self.params.get("run_db_node_benchmarks"):
+            self.log.info("Validate node benchmarks results")
+            compare_results = self.db_cluster.get_node_benchmarks_results() or {}
+            ready_nodes = []
+            for node, items in compare_results.items():
+                for item, result in items.items():
+                    ready_nodes.append(result["is_within_margin"])
+                    if not result["is_within_margin"]:
+                        self.log.error("HW performance test on node %s has bad results for %s : %s", node, item, result)
+            if not all(ready_nodes):
```
Comment (roydahan) [inline performance_regression_test.py:62]: Let's add a new parameter that decides if to stop the test if above margin or not.

File: sdcm/utils/benchmarks.py [hunk 3]
```diff
@@ -129,25 +131,53 @@ def _collect_benchmark_output(self):
...
+            return []
+        return [doc["_source"] for doc in results["hits"]["hits"]]
 
     def _compare_results(self):
+        if not self._benchmark_runners:
+            return
+
+        # assume that cluster nodes have same instance type
+        instance_type = self._benchmark_runners[0].node_instance_type
+        averages = self._get_average_results(es_docs=self._get_benchmark_results(instance_type))
```
Comment (roydahan) [inline sdcm/utils/benchmarks.py:179]: Let's report these results also to the performance email.

File: jenkins-pipelines/perf-regression-throughput-30min-with-hw-checks.jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#!groovy
+
+// trick from https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/43
+def lib = library identifier: 'sct@snapshot', retriever: legacySCM(scm)
+
+perfRegressionParallelPipeline(
+    backend: "aws",
+    region: "us-east-1",
+    test_name: "performance_regression_test.PerformanceRegressionTest",
+    test_config: '''["test-cases/performance/perf-regression.100threads.30M-keys.yaml","configurations/run_db_node_benchmarks.yaml"]''',
```
Comment (roydahan) [inline jenkins-pipelines/perf-regression-throughput-30min-with-hw-checks.jenkinsfile:10]: Let's add this also to the performance_with_nemesis test.

**Review by roydahan — CHANGES_REQUESTED** [review]
Everything looks good.
I left few comments to address, then we can merge.

**Review by KnifeyMoloko — APPROVED** [review]
Approved

File: performance_regression_test.py (hunk 2 above)
Comment (aleksbykov) [reply performance_regression_test.py:62]: Done

File: jenkins-pipelines/perf-regression-throughput-30min-with-hw-checks.jenkinsfile (hunk 4 above)
Comment (aleksbykov) [reply jenkins-pipelines/perf-regression-throughput-30min-with-hw-checks.jenkinsfile:10]: done

File: performance_regression_test.py [hunk 5]
```diff
@@ -46,7 +46,26 @@ def __init__(self, *args):
...
+    @log_run_info
+    def setUp(self):
+        super().setUp()
+        if self.params.get("run_db_node_benchmarks"):
+            self.log.info("Validate node benchmarks results")
+            compare_results = self.db_cluster.get_node_benchmarks_results() or {}
+            ready_nodes = []
+            for node, items in compare_results.items():
+                for item, result in items.items():
+                    ready_nodes.append(result["is_within_margin"])
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply performance_regression_test.py:60]: done

**Comment by aleksbykov** [comment]
@roydahan @KnifeyMoloko Could you review one more time. Added results to report, add 2 new events, added new flag to control test termination.

**Comment by aleksbykov** [comment]
Screenshot from 2022-12-28 14-21-49

Example of email report with node benchmark results

**Review by KnifeyMoloko — APPROVED** [review]
LGTM (apart from the merge conflict in sct_config.py)

**Comment by aleksbykov** [comment]
@roydahan , conflicts have been resolved.

File: jenkins-pipelines/perf-regression-latency-250gb-with-nemesis-with-hw-checks.jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!groovy
```
Comment (roydahan) [inline jenkins-pipelines/perf-regression-latency-250gb-with-nemesis-with-hw-checks.jenkinsfile:1]: @aleksbykov I don't think it should be a new pipeline, we should add it to all official pipelines of performance regressions we are switching to.

File: jenkins-pipelines/perf-regression-throughput-30min-with-hw-checks.jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#!groovy
```
Comment (roydahan) [inline jenkins-pipelines/perf-regression-throughput-30min-with-hw-checks.jenkinsfile:1]: Same as above.

File: sdcm/sct_config.py [hunk 8]
```diff
@@ -1426,6 +1426,9 @@ class SCTConfiguration(dict):
              help="""add/remove num of process on each round"""),
         dict(name="use_hdr_cs_histogram", env="SCT_USE_HDR_CS_HISTOGRAM", type=boolean,
              help="""Enable hdr histogram logging for cs"""),
+
+        dict(name="stop_on_hw_perf_failure", env="SCT_STOP_ON_HW_PERF_FAILURE", type=boolean,
+             help="""Stop performance test if hardware performance test failed""")
```
Comment (roydahan) [inline sdcm/sct_config.py:1431]: The description isn't good enough. please explain what is considered as failed, if it's within margin and what's the margin and refer to the relevant params.

**Review by roydahan — CHANGES_REQUESTED** [review]

File: jenkins-pipelines/perf-regression-throughput-30min-with-hw-checks.jenkinsfile (hunk 7 above)
Comment (aleksbykov) [reply jenkins-pipelines/perf-regression-throughput-30min-with-hw-checks.jenkinsfile:1]: done

File: sdcm/sct_config.py (hunk 8 above)
Comment (aleksbykov) [reply sdcm/sct_config.py]: Update description

**Comment by aleksbykov** [comment]
@roydahan could you review it?

**Review by roydahan — APPROVED** [review]

**Pull request merged**